tidy3d/components/time_modulation.py

```diff
@@ -0,0 +1,207 @@
+""" Defines time modulation to the medium"""
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import Union
+from math import isclose
+
+import pydantic.v1 as pd
+import numpy as np
+
+from .base import Tidy3dBaseModel, cached_property
+from .types import InterpMethod
+from .source import AbstractTimeDependence
+from .data.data_array import SpatialDataArray
+from ..exceptions import ValidationError
+from ..constants import HERTZ
+
+
+class AbstractTimeModulation(AbstractTimeDependence):
+    """Base class for modulation in time.
+
+    Note
+    ----
+    This class describes the time dependence part of the separable space-time modulation type
+    as shown below,
+    .. math::
+
+        amp(r, t) = \\Re[amp\\_time(t) \\cdot amp\\_space(r)]
+
+    """
+
+    @cached_property
+    @abstractmethod
+    def max_modulation(self) -> float:
+        """Estimated maximal modulation amplitude."""
+
+
+class ContinuousWaveTimeModulation(AbstractTimeDependence):
+    """The modulation in time is in continuous oscillation.
+
+    Note
+    ----
+    .. math::
+
+        amp\\_time(t) = amplitude \\cdot \\
+                e^{i \\cdot phase - 2 \\pi i \\cdot freq0 \\cdot t}
+
+    Note
+    ----
+    The full space-time modulation is,
+    .. math::
+
+        amp(r, t) = \\Re[amp\\_time(t) \\cdot amp\\_space(r)]
+
+
+    Example
+    -------
+    >>> cw = ContinuousWaveTimeModulation(freq0=200e12, amplitude=1, phase=0)
+    """
+
+    freq0: pd.PositiveFloat = pd.Field(
+        ..., title="Modulation Frequency", description="Modulation frequency.", units=HERTZ
+    )
+
+    def amp_time(self, time: float) -> complex:
+        """Complex-valued source amplitude as a function of time."""
+
+        omega = 2 * np.pi * self.freq0
+        return self.amplitude * np.exp(-1j * omega * time + 1j * self.phase)
+
+    @cached_property
+    def max_modulation(self) -> float:
+        """Estimated maximal modulation amplitude."""
+        return abs(self.amplitude)
+
+
+TimeModulationType = Union[ContinuousWaveTimeModulation]
+
+
+class AbstractSpaceModulation(ABC, Tidy3dBaseModel):
+    """Base class for modulation in space.
+
+    Note
+    ----
+    This class describes the 2nd term in the full space-time modulation below,
+    .. math::
+
+        amp(r, t) = \\Re[amp\\_time(t) \\cdot amp\\_space(r)]
+
+    """
+
+    @cached_property
+    @abstractmethod
+    def max_modulation(self) -> float:
+        """Estimated maximal modulation amplitude."""
+
+
+class SpaceModulation(AbstractSpaceModulation):
```

Comment:
tempted to call this `SpatialModulation` for consistency with `SpatialDataArray`. Not sure, any thoughts?

Reply:
But we also have SpaceTimeModulation. SpaceModulation is more consistent with that?

Reply:
yea that's true, we can leave it.